src/internal/packager/images/push.go

```diff
@@ -68,53 +49,98 @@ func Push(ctx context.Context, cfg PushConfig) error {
 		return fmt.Errorf("failed to instantiate oci directory: %w", err)
 	}
 
-	pushImage := func(srcName, dstName string) error {
-		remoteRepo := &orasRemote.Repository{
-			PlainHTTP: plainHTTP,
-			Client:    client,
-		}
-		remoteRepo.Reference, err = registry.ParseReference(dstName)
-		if err != nil {
-			return fmt.Errorf("failed to parse ref %s: %w", dstName, err)
-		}
-		defaultPlatform := &ocispec.Platform{
-			Architecture: cfg.Arch,
-			OS:           "linux",
+	// The user may or may not have a cluster available, if it's available then use it to connect to the registry
+	c, _ := cluster.NewCluster()
+	err = retry.Do(func() error {
+		// Include tunnel connection in case the port forward breaks, for example, a registry pod could spin down / restart
+		var tunnel *cluster.Tunnel
+		if c != nil {
+			var err error
+			registryURL, tunnel, err = c.ConnectToZarfRegistryEndpoint(ctx, cfg.RegistryInfo)
+			if err != nil {
+				return err
+			}
+			if tunnel != nil {
+				defer tunnel.Close()
+			}
 		}
-		if tunnel != nil {
-			return tunnel.Wrap(func() error {
-				return copyImage(ctx, src, remoteRepo, srcName, dstName, cfg.OCIConcurrency, defaultPlatform)
-			})
+		client := &auth.Client{
+			Client: orasRetry.DefaultClient,
+			Cache:  auth.NewCache(),
+			Credential: auth.StaticCredential(registryURL, auth.Credential{
+				Username: cfg.RegistryInfo.PushUsername,
+				Password: cfg.RegistryInfo.PushPassword,
+			}),
 		}
-		return copyImage(ctx, src, remoteRepo, srcName, dstName, cfg.OCIConcurrency, defaultPlatform)
-	}
 
-	for _, img := range cfg.ImageList {
-		l.Info("pushing image", "name", img.Reference)
-		// If this is not a no checksum image push it for use with the Zarf agent
-		if !cfg.NoChecksum {
-			offlineNameCRC, err := transform.ImageTransformHost(registryURL, img.Reference)
+		plainHTTP := cfg.PlainHTTP
+
+		if dns.IsLocalhost(registryURL) && !cfg.PlainHTTP {
+			var err error
+			plainHTTP, err = shouldUsePlainHTTP(ctx, registryURL, client)
 			if err != nil {
 				return err
 			}
+		}
 
-			if err = pushImage(img.Reference, offlineNameCRC); err != nil {
-				return err
+		pushImage := func(srcName, dstName string) error {
```

Comment:
Declaring this as a private package func is also a valid approach, and a good way to test how portable logic in your function is (e.g. how easy it is to change should requirements change in the future). However, the params get much more complicated without using the parent func's scope, so this feels like a reasonable tradeoff here. IMO this is about the upper limit of how complex a locally scoped function variable should be.